ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedPfxStore.java

```diff
@@ -0,0 +1,241 @@
+        Extension.basicConstraints, true, new BasicConstraints(IS_CA));
+    v3CertificateBuilder.addExtension(
+        Extension.subjectAlternativeName, false, getSubjectAlternativeNames());
+
+    final ContentSigner contentSigner =
+        new JcaContentSignerBuilder("SHA256WithRSAEncryption").build(keyPair.getPrivate());
+
+    this.certificate =
+        new JcaX509CertificateConverter()
+            .setProvider(BOUNCY_CASTLE_PROVIDER)
+            .getCertificate(v3CertificateBuilder.build(contentSigner));
+  }
+
+  private GeneralNames getSubjectAlternativeNames() {
+    final List<GeneralName> hostGeneralNames =
+        sanHostNames.stream()
+            .map(hostName -> new GeneralName(GeneralName.dNSName, hostName))
+            .collect(Collectors.toList());
+    final List<GeneralName> ipGeneralNames =
+        sanIpAddresses.stream()
+            .map(ipAddress -> new GeneralName(GeneralName.iPAddress, ipAddress))
+            .collect(Collectors.toList());
+    final GeneralName[] generalNames =
+        Stream.of(hostGeneralNames, ipGeneralNames)
+            .flatMap(Collection::stream)
+            .toArray(GeneralName[]::new);
+
+    return new GeneralNames(generalNames);
+  }
+
+  private void createKeyStore() throws IOException, GeneralSecurityException {
+    this.keyStore = convert(true);
+  }
+
+  private void createTrustStore() throws IOException, GeneralSecurityException {
+    this.trustStore = convert(false);
+  }
+
+  private Path convert(final boolean isKeyStore) throws IOException, GeneralSecurityException {
+    final KeyStore keyStore = KeyStore.getInstance("PKCS12");
+    keyStore.load(null);
+    if (isKeyStore) {
+      keyStore.setKeyEntry(alias, keyPair.getPrivate(), password, new Certificate[] {certificate});
+    } else {
+      keyStore.setCertificateEntry(alias, certificate);
+    }
+
+    return saveKeyStore(keyStore);
+  }
+
+  private Path saveKeyStore(final KeyStore keyStore)
+      throws IOException, KeyStoreException, NoSuchAlgorithmException, CertificateException {
+    final Path pfxPath = Files.createTempFile(parentPath, alias, ".pfx");
+    pfxPath.toFile().deleteOnExit();
+    try (final FileOutputStream outputStream = new FileOutputStream(pfxPath.toFile())) {
+      keyStore.store(outputStream, password);
+    }
+    return pfxPath;
+  }
+
+  private void createKnownClientsFile() throws IOException, CertificateEncodingException {
+    final RDN commonNameRDN = new X500Name(distinguishedName).getRDNs(BCStyle.CN)[0];
+    final String commonName = IETFUtils.valueToString(commonNameRDN.getFirst().getValue());
+    final String fingerPrint = TLS.certificateHexFingerprint(certificate);
+
+    final Path tempFile = Files.createTempFile(parentPath, alias + "knownClientsFile", ".txt");
+    tempFile.toFile().deleteOnExit();
+    Files.writeString(tempFile, commonName + " " + fingerPrint);
+    this.knownClientsFile = tempFile;
+  }
+
+  private void createPasswordFile() throws IOException {
```

Comment:
nit: does a certificate need a password _file_ - or just a password? Regarding abstractions, is it useful to get this key's password, and let the user of this class write it to file?

Reply:
Refactored.